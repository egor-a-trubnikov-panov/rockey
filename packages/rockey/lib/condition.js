import interpolateString from './utils/interpolateString';

const condition = conditionFunc => (...args) => {
  if (conditionFunc()) {
    return interpolateString(...args);
  } else {
    return null;
  }
};

export default condition;
