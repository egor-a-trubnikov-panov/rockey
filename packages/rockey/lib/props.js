import when from './when';

const props = name => {
  return when(name, () => true);
};

export default props;
