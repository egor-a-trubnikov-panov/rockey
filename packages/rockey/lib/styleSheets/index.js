import StyleSheet from './StyleSheet';

const sheet = new StyleSheet();

let insertWithQueue = false;

export const insert = precss => {
  if (insertWithQueue) {
    sheet.insertWithQueue(precss);
  } else {
    sheet.insert(precss);
  }
};

export const getRules = css => sheet.getRules();
export const clearStyles = css => sheet.clear();

export const speedy = () => sheet.enableSpeedy();
export const useQueue = () => {
  insertWithQueue = true;
};

export default insert;
