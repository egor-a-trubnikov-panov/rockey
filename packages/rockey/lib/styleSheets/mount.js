import StyleSheet from './StyleSheet';

const isBrowser = typeof document !== 'undefined';

const createStyleSheet = type => {
  if (!isBrowser) {
    // TODO: add support to server rendering
    return {
      children: [],
      type,
      innerHtml: '',
      textContent: '',
      appendChild: node => {},
    };
  } else {
    const mountNode = document.createElement('style');

    mountNode.type = 'text/css';

    const link = document.querySelector('link[rel="stylesheet"]');
    if (link) {
      document.head.insertBefore(mountNode, link);
    } else {
      document.head.appendChild(mountNode);
    }

    mountNode.textContent = '';

    return mountNode;
  }
};

const mount = () => {
  return new StyleSheet({
    rules: createStyleSheet('rules'),
    mixins: createStyleSheet('mixins'),
  });
};

export default mount;
