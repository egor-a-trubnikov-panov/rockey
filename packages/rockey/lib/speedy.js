import { speedy } from './styleSheets';
speedy();
