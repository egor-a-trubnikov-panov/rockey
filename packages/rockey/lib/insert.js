import { staticRule } from './rule';

export const insert = (strings, ...values) => {
  const css = staticRule(strings, ...values);
  css.getClassList();
};

export default insert;
