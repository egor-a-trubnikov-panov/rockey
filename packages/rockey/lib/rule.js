import createParser from 'rockey-css-parse';
import isString from 'lodash/isString';
import * as classnames from 'rockey-css-parse/utils/classnames';

import {
  getClassName,
  getAnimationName,
  getMixinClassName,
} from './css/getClassName';

import insert from './styleSheets';
import wrapTemplateStrings from './utils/wrapTemplateStrings';

import vendorPrefix from './plugins/vendorPrefix';
import validateCSSRule from './plugins/validateCSSRule';

const plugins = [];

plugins.push(vendorPrefix());

// if (process.env.NODE_ENV !== 'production') {
//   plugins.push(validateCSSRule());
// }

const parse = createParser({
  getClassName,
  getMixinClassName,
  getAnimationName,
  plugins,
});

const getSelector = classnames.getSelector(getClassName);

export default function rule(raw, ...values) {
  // export default function rule(...args) {
  let parent = null;
  let cached = null;
  let inserted = false;
  let isAlreadyParsed = false;
  let addedMixins = [];

  if (Array.isArray(raw)) {
    isAlreadyParsed = !isString(raw[0]);
  } else if (isString(raw)) {
  } else {
    throw new Error('rockey-rule: wrong call type');
  }

  const key = JSON.stringify({
    raw,
    values,
  });

  return {
    wrapWith(displayName) {
      if (cached) {
        throw new Error(
          'Can not wrap because rule is already parsed and inserted'
        );
      }

      if (isAlreadyParsed) {
        const selector = getSelector(displayName);

        raw.forEach(precss => {
          precss.selector = precss.selector.map(p => `${selector} ${p}`);
          precss.root = [displayName];

          precss.mixins.forEach(m => {
            m.updateSelector(ms => `${selector} ${ms}`);
          });
        });
      } else {
        raw = wrapTemplateStrings(displayName, raw);
      }
    },

    addParent(rule) {
      parent = rule;
    },

    parse() {
      let precss = null;
      let classList = null;

      if (!cached) {
        if (isAlreadyParsed) {
          precss = raw;
          classList = {};

          raw.forEach(part => {
            part.root.forEach(c => {
              if (!classList[c]) {
                classList[c] = getClassName(c);
              }
            });
          });
        } else {
          let parsed = parse(raw, ...values);

          precss = parsed.precss;
          classList = parsed.classList;
        }

        if (addedMixins.length) {
          Object.keys(classList).forEach(displayName => {
            addedMixins.forEach(addedMixin => {
              const css = rule`
                ${displayName} {
                  ${addedMixin}
                }
              `;

              css.parse().precss.forEach(p => {
                precss.push(p);
              });

              // precss = precss.concat(css.parse().precss);
            });
          });
        }

        cached = {
          precss,
          classList,
        };
      } else {
        precss = cached.precss;
        classList = cached.classList;
      }

      return { precss, classList };
    },

    transform(transformFunc) {
      let precss = {};

      if (isAlreadyParsed) {
        precss = raw;
      } else {
        const parsed = this.parse();
        precss = parsed.precss;
      }

      const tree = {};

      precss.forEach(part => {
        part.root.forEach(key => {
          if (!tree[key]) {
            tree[key] = [];
          }
          tree[key].push(part);
        });
      });

      return transformFunc(tree, transformed => {
        return rule(Array.isArray(transformed) ? transformed : [transformed]);
      });
    },

    addMixins(mixins) {
      addedMixins = addedMixins.concat(mixins);
    },

    getClassList(props = {}) {
      let parentClasses = [];

      if (parent) {
        const parentClassList = parent.getClassList(props);
        Object.keys(parentClassList).forEach(name => {
          parentClasses.push(parentClassList[name]);
        });
      }

      const { classList, precss } = this.parse();

      if (!inserted) {
        if (precss.length) {
          insert(precss);
        }

        inserted = true;
      }

      let resultClassList = {};

      if (parent) {
        let parentClassList = parentClasses.join(' ');
        if (parentClassList) {
          Object.keys(classList).forEach(name => {
            resultClassList[name] = `${classList[name]} ${parentClassList}`;
          });
        }
      } else {
        resultClassList = Object.assign({}, classList);
      }

      let mixinsPrecss = [];

      precss.forEach(pre => {
        if (pre.mixins) {
          pre.mixins.forEach(mixin => {
            const result = mixin(props);

            if (result) {
              if (result.className) {
                pre.root.forEach(c => {
                  resultClassList[c] += ` ${result.className}`;
                });
              }

              if (result.precss) {
                result.precss.forEach(p => {
                  mixinsPrecss.push(p);
                });
              }
            }
          });
        }
      });

      if (mixinsPrecss.length) {
        insert(mixinsPrecss);
      }

      return resultClassList;
    },
  };
}
