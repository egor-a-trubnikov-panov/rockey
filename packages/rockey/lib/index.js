import condition from './condition';
import rule from './rule';
import props from './props';
import insert from './insert';
import when from './when';

const rockey = {
  condition,
  rule,
  insert,
  when,
};

export default rockey;
export { condition };
export { insert };
export { when };
export { rule };
export { props };
