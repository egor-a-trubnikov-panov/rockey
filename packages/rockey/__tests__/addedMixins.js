import rule from '../lib/rule';

import * as styleSheetsModule from '../lib/styleSheets';

jest.mock('../lib/utils/hash', () => {
  return () => 'hash';
});

const insert = jest.fn();
styleSheetsModule.default = insert;

test('rule', () => {
  const css = rule`
    Button {
      color: red;
    }
  `;

  css.addMixins([props => (props.primary ? 'color:blue' : 'color:red')]);

  const classList = css.getClassList({
    primary: true,
  });

  expect(classList).toMatchSnapshot();
  expect(insert.mock.calls).toMatchSnapshot();

  const classList2 = css.getClassList({
    primary: false,
  });

  expect(classList2).toMatchSnapshot();
  expect(insert.mock.calls).toMatchSnapshot();
});
