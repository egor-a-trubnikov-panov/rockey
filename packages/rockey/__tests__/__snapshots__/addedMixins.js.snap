// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rule 1`] = `
Object {
  "Button": "Button-hash m-anon-hash-1",
}
`;

exports[`rule 2`] = `
Array [
  Array [
    Array [
      Object {
        "media": null,
        "mixins": Array [],
        "root": Array [
          "Button",
        ],
        "selector": Array [
          ".Button-hash",
        ],
        "styles": Object {
          "color": "red",
        },
      },
      Object {
        "media": null,
        "mixins": Array [
          [Function],
        ],
        "root": Array [
          "Button",
        ],
        "selector": Array [
          ".Button-hash",
        ],
        "styles": Object {},
      },
    ],
  ],
  Array [
    Array [
      Object {
        "frames": undefined,
        "selector": Array [
          ".m-anon-hash-1.Button-hash",
        ],
        "styles": Object {
          "color": "blue",
        },
      },
    ],
  ],
]
`;

exports[`rule 3`] = `
Object {
  "Button": "Button-hash m-anon-hash-2",
}
`;

exports[`rule 4`] = `
Array [
  Array [
    Array [
      Object {
        "media": null,
        "mixins": Array [],
        "root": Array [
          "Button",
        ],
        "selector": Array [
          ".Button-hash",
        ],
        "styles": Object {
          "color": "red",
        },
      },
      Object {
        "media": null,
        "mixins": Array [
          [Function],
        ],
        "root": Array [
          "Button",
        ],
        "selector": Array [
          ".Button-hash",
        ],
        "styles": Object {},
      },
    ],
  ],
  Array [
    Array [
      Object {
        "frames": undefined,
        "selector": Array [
          ".m-anon-hash-1.Button-hash",
        ],
        "styles": Object {
          "color": "blue",
        },
      },
    ],
  ],
  Array [
    Array [
      Object {
        "frames": undefined,
        "selector": Array [
          ".m-anon-hash-2.Button-hash",
        ],
        "styles": Object {
          "color": "red",
        },
      },
    ],
  ],
]
`;
